File: frontend/gameapp-frontend/src/application/repositories/iitem-repository.ts
import type { ItemType } from '../enumerates/item-type';
import { Item } from '../../domain/entities/items/item';
import { ItemName } from '../../domain/value-objects/items/item-name';
import type { ItemRarity } from '../../domain/enumerates/item-rarity';

export interface IItemRepository {
    fetchById(id: string): Promise<Item>;
    fetchByName(name: ItemName): Promise<Item>;
    fetchAll(): Promise<Item[]>;
    fetchAllByFilter(type?: ItemType, rarity?: ItemRarity): Promise<Item[]>;
    save(item: Item): Promise<Item>;
    delete(id: string): Promise<Item>;
    update(id: string, item: Item): Promise<Item>;
}

File: backend/GameApp.Application/Repositories/IItemRepository.cs
using GameApp.Application.Enumerates;
using GameApp.Domain.Entities.Items;
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.Repositories;

public interface IItemRepository
{
    public abstract Task<IEnumerable<Item>> FetchAllAsync();

    public abstract Task<IEnumerable<Item>> FetchAllByFilterAsync(ItemType? type, ItemRarity? rarity);

    public abstract Task<Item?> FetchByIdAsync(Guid id);

    public abstract Task<Item?> FetchByName(ItemName name);

    public abstract Task<Item?> DeleteAsync(Guid id);
    
    public abstract Task<Item?> SaveAsync(Item Item);

    public abstract Task<Item?> UpdateAsync(Guid id, Item Item);

    public abstract Task SeedAsync();
}
